Restructure and update community governance docs

AevaOnline opened the pull request:
This is a first pass at re-organizing the community documentation. I anticipate a healthy conversation and several iterations of this work before it is acceptable to merge. It's a hefty patch, with substantial implications for the project, and I'd like folks to understand what I am suggesting before we merge this.

Changes:
- simplifying front page / README
- creating a docs/Community folder for documentation about community structure/governance
- refactored the top-level README to be a succinct summary of the project,
  with content mostly moved to docs/GettingStartedDocs/README
- added docs/Community/README, which outlines the community structure and helpfully links to other documents
- moved docs/Contributing to docs/Community/Contributing, clarified the
  "Design Discussion" section and added a "Weekly Issue Triage" section
- refactored docs/Governance into
    docs/Community/Governance  (which is also significantly expanded upon)
    docs/Community/ReleaseProcess  (mostly unchanged)
    docs/Community/conduct/ConflictPolicy  (mostly unchanged)
- refactored docs/Committers into
    docs/Community/Committers  (retains list of committers, though I'd like to remove this)
    docs/Community/governance/charter  (retains definition of scope)
    docs/Community/governance/README  (retains list of committee members)
- added a few SIG folders and templates as placeholders and examples
- added a few other files as placeholders for content that would be helpful to write soon

To get past the license check affecting documentation, I also added a
wildcard match in scripts.check-license.ignore for docs/*

Signed-off-by: Aeva Black <806320+AevaOnline@users.noreply.github.com>

achamayou commented on README.md:
```diff
@@ -8,79 +8,47 @@ Open Enclave SDK
-- [Windows Server 2016 with SGX hardware](docs/GettingStartedDocs/install_oe_sdk-Windows.md)
+The Open Enclave SDK is a hardware-agnostic open source library for developing
+applications that run inside Hardware Trusted Execution Environments, also known
+as Enclaves.
```
This is a slight change in direction from the previous stated goal, which was to target Trusted Execution Environments "usually secured by hardware" but possibly also including software solutions like https://docs.microsoft.com/en-us/windows-hardware/design/device-experiences/oem-vbs, which aren't.

I have no strong opinion on the change itself, but just wanted to confirm this is deliberate.

AevaOnline replied:
Great callout. Yes, it was deliberately to align this project with the CCC's updated charter - 
https://github.com/confidential-computing/governance/blob/master/scoping.md#executive-summary

achamayou replied:
Ok, thank you for the link!

achamayou commented on README.md:
```diff
@@ -8,79 +8,47 @@ Open Enclave SDK
 
-This project follows a [Code of Conduct](docs/CodeOfConduct.md) adapted from the
-[Contributor Covenant v1.4](https://www.contributor-covenant.org).
+You'll find compresenive documentation in the
```
```suggestion
You'll find comprehensive documentation in the
```

achamayou commented on docs/Community/CodeOfConduct.md:
```diff
@@ -0,0 +1,11 @@
+# Open Enclave SDK Code of Conduct
```
I would like to suggest that we use one of the standard locations/filenames GitHub expects: https://help.github.com/en/github/building-a-strong-community/adding-a-code-of-conduct-to-your-project#adding-a-code-of-conduct-manually

That way it will get displayed in the Community tab, and it will probably be slightly easier to find.

AevaOnline replied:
Great catch! I like it and will update. Thanks.

achamayou commented on docs/Community/Committers.md:
```diff
@@ -0,0 +1,50 @@
+Desired Status
+--------------
+
+As the project grows, in order to scale the management of the project, it is becoming increasingly necessary to delegate responsibility to domain-owners and improve the community's development workflow through automation. This will be handled through the use of [Prow](https://git.k8s.io/test-infra/prow).
```
Is prow going to replace bors?

AevaOnline replied:
That would be my preference. It is substantially more feature rich, and supports many larger projects (ie, will support OE SDK as we grow). 

As far as I'm aware, bors does not support the workflows that I recommend we adopt.

achamayou commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+The Open Enclave SDK community uses a
+[concensus-seeking](https://en.wikipedia.org/wiki/Consensus-seeking_decision-making)
+approach to governance at all levels. To support a growing project, we have
+adopted a loose heirarchy wherein domain-specific decisions are handled within
```
```suggestion
adopted a loose hierarchy wherein domain-specific decisions are handled within
```

achamayou commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+through Working Groups (**WG**), and a Community Governance Committee (**CGC**)
+meets to review the overall structure, ensuring the health of the project.
+
+Each group (SIGs, WGs, and CGC) maintain documentation in this repository.
```
```suggestion
Each group (SIGs, WGs, and CGC) maintains documentation in this repository.
```

achamayou commented on docs/Community/sig-architecture/README.md:
```diff
@@ -0,0 +1,26 @@
+
+SIG-Architecture is responsible for reviewing high level designs, which inform project goals and technical norms.
+
+This SIG is responsible for all project [DesignDocs](../../DesignDocs/README.md) and should seek concensus with the SIG Leads for any design proposal before approving the document.
```
```suggestion
This SIG is responsible for all project [DesignDocs](../../DesignDocs/README.md) and should seek consensus with the SIG Leads for any design proposal before approving the document.
```

achamayou commented on docs/Community/triage/README.md:
```diff
@@ -0,0 +1,31 @@
+
+    Triage is the process of determining the priority of patients' treatments
+    based on the severity of their condition or likelihood of recovery with and
+    without treatment.
```
I suggest pointing to https://en.wikipedia.org/wiki/Software_bug#Bug_management instead. Although I understand this may be where the term originates, I'm not sure it will be helpful to someone who hasn't come across the term before.

AevaOnline replied:
The word "triage" is nowhere to be found on that page, though it is a better description of what we do in the triage meeting.

I think it is useful, especially for folks who have not come across the term "triage" before (eg, non-native english speakers) to understand the context it conveys, and I also agree that providing a link to deeper description of Bug Management is helpful. Thanks.

achamayou replied:
Yes, you're right, I got there through https://en.wikipedia.org/wiki/Triage_(disambiguation) and "Bug Triage", but I see the landing location doesn't mention the word, so it's not that helpful.

dthaler commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+considered authoritative for that subsection of the project's code base insofar
+as code reviews and architectural discussions. When a change affects code owned
+by multiple SIGs, it should be coordinated with the respective SIG leads.
+Cross-SIG efforts can be coordinated by establishing a Working Group (WG), which
```
Q: Is a WG needed for every cross-SIG effort? (I think no) so would be helpful to know when extra process is needed vs not needed.  (Lightweight is good so spinning up a WG for one PR would be overkill.)
Right now, I'm hesitant to add "WG" discussion in here until there's actually shown to be a need for one, but open to discussion.

AevaOnline replied:
I thought about omitting WGs from the doc, as they do feel a little process-heavy right now, but on the other hand, I anticipate a need for work that spans several SIGs pretty soon after we adopt a structure that only has SIGs and Committees.

I could break the WG component into a separate PR so that it can sit in the backlog until we need it?

dthaler replied:
I like that idea.  I also have no objection to including it now if others think it's good to document it even if we don't need it yet.

achamayou replied:
I'm in favour of keeping it there, since the description is quite lightweight. It's possible that as the first WG is formed, this may get refined, but that can happen in due course.

AevaOnline replied:
:+1:

achamayou commented on docs/Community/triage/README.md:
```diff
@@ -0,0 +1,31 @@
+
+Meetings
+--------
+
```
```suggestion
Triage meetings are open to everyone. If you have recently filed an issue or a PR, you may want to attend the next triage meeting and help with classification.
```

dthaler commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+* on our mailing list or chat server (details TBD) (TODO)
+
+While we strive for openness and transparency, it is natural that some
+discussions will also happen in person (eg, in the hallway track) or other
```
"eg," -> "e.g.,"

but what is "the hallway track" referring to?   OE doesn't have f2f meetings today or "tracks".

jhand2 replied:
I think this is just using the conference term "hallway track" as an analogy. Some people that work on the SDK inevitably work in close physical proximity and not all discussions will happen online. Given that reality, I think this is good guidance in terms of decision-making for the SDK.

This is language that is pretty specific to people who attend a lot of conferences (I hear it most often used by academics) so I agree that language could be a little confusing.

AevaOnline replied:
Jordan, your interpretation is correct. Even without a conference / event, before COVID there were random f2f discussions that happened between some folks who work in the same office. This is human nature and not something I want to discourage -- it is, rather, something I want to include and support in the process by clearly asking folks to share the outcome of any such discussions.

That said, I thank you for calling out language that isn't inclusive. I try, but still sometimes forget that I'm speaking in culturally-specific jargon some times.

I'll add clarify my use of the phrase "hallway track"

dthaler commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+The community strives to triage and respond to all incoming issues within one
+week. If you haven't heard anything by then, or an issue or PR is not getting
+any response within a resonable time after it is triaged, please feel free to
+remind us with a ping on the thread.
```
Change "thread" to "issue or PR"

AevaOnline replied:
Is this critique valuable?

Isn't a discussion on either a PR or an issue a "thread"?

dthaler commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+
+The community strives to triage and respond to all incoming issues within one
+week. If you haven't heard anything by then, or an issue or PR is not getting
+any response within a resonable time after it is triaged, please feel free to
```
typo: resonable

dthaler commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+All contributions are expected to be reviewed by at least one SIG lead other
+than the contributor, and have no significant objections from other SIG leads
+within the same SIG. Objections from leads of other SIGs should be respected,
+and consensus sought (but not required); overruling the objection of another SIG
```
If a PR affects multiple SIGs then I think this text in 128-129 is incorrect since it implies that only one SIG is required and the other is not required,.

AevaOnline replied:
My intent is to distinguish between two scenarios:
1. a PR affects code owned by two SIGs (eg, code in two different directories). It therefore requires approval from both SIGs
2. a PR only affects code owned by one SIG, but members of another SIG have strong opinions (objections) to it (eg, because of implications to their domain, or project-wide impact,... or just a disagreement over style or use of punctuation). In this case, it is still the final decision of just one SIG, though if serious conflict can not be resolved between the SIGs, it should be brought to the CGC.

dthaler replied:
Completely agree with that.  Any way to make that be clearer in the text?

AevaOnline replied:
I'll give it a try

dthaler commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+lead should only be done after careful consideration.
+
+All contributions must also satisfy the
+[contribution guidelines](Contributing.md), such as having a DCO and being
```
undefined acronym.
replace DCO with [Developer Certificate of Origin (DCO)](https://developercertificate.org/)

dthaler commented on docs/Community/ReleaseProcess.md:
```diff
@@ -0,0 +1,51 @@
+
+We want to ensure that our releases go through a process of community feedback.
+That means that before any release is finalized, a feedback solicitation period
+will happen. The release manager will open a GitHub issue announcing a Release
```
the term "release manager" is used here without reference or definition.
I have previously asked "Who is the release manager and how are they selected"
I *think* from your PR that the answer in your terminology might be "the lead of the Release SIG" in which case we should consider using that term instead

AevaOnline replied:
This section is simply copied from previous documentation.

radhikaj replied:
We previously voted on the release manager in CGC. The release manager is currently me(after John left)

dthaler replied:
@radhikaj Agreed, I'm imagining that we label it as a Release SIG with you as the SIG lead, and any public release meeting you run would be a Release SIG meeting.  Then we don't need to treat "voting on a release manager" as any separate process if you step down in the future, it'd be the same as for any other SIG.

dthaler commented on docs/Community/ReleaseProcess.md:
```diff
@@ -0,0 +1,51 @@
+close the current RC issue and open a new one to begin the process again.
+
+We do not intend to finalize a release until the majority of the community
+agrees it is ready. Once a consensus is reached, that is, all major grievances
```
passive tense in "consensus is reached" means subject is missing.
Is it a consensus among the Release SIG, or of some other body responsible for determining/ensuring consensus?

AevaOnline replied:
this was simply copied from elsewhere.

dthaler commented on docs/Community/conduct/ConflictPolicy.md:
```diff
@@ -0,0 +1,26 @@
+threatening or harassing, report it. We are dedicated to providing an environment
+where participants feel welcome and safe.
+
+Reports can be made to any member of the [Community Governance Committee](
```
In your Community/CodeOfConduct.md lines 8-11 say: "Instances of abusive, harassing, or otherwise unacceptable behavior may be reported to the Open Enclave [Community Governance Committee](governance/README.md) or to the [Confidential Computing Consortium](https://github.com/confidential-computing/governance/blob/master/code-of-conduct.md)."

But here only mentions the former.  We should be consistent, whichever we use.

AevaOnline replied:
I disagree that this needs to be so formal and consistent. It's community documentation, not a C header file.

dthaler commented on docs/Community/sig-attestation/README.md:
```diff
@@ -0,0 +1,26 @@
+==================================
+
+SIG-Attestation is responsible for code, plugins, testing, and other works
+related to or required for in-enclave attestation and integration with
```
remove "in-enclave" since there is an experimental "host verification" library for verifying attestation reports outside of an enclave.

AevaOnline replied:
Oh, neat!

dthaler commented on docs/Community/sig-testing/README.md:
```diff
@@ -0,0 +1,35 @@
+TBD
+
+Charter
+-------
```
should there be a TBD under this line?

AevaOnline replied:
Probably. But this is just a copy-pasta of a template, to demonstrate what we might put here.

I'd like to see these files filled in through community discussion  :)

dthaler commented on docs/GettingStartedDocs/README.md:
```diff
@@ -0,0 +1,93 @@
+without an enclave (**TODO: add link to docs for this**), or you can use one of
+the hardware types below.
+
+**TODO**: start separating documentation by hardware type
```
Not sure what this means since SGX and OP-TEE have separate docs now.  (There are certainly many issues with the docs but I don't yet follow what this TODO is.)

AevaOnline replied:
I haven't read through all the docs yet, so didn't I realize that. thanks for spotting this. I'll remove this TODO.

AevaOnline replied:
```suggestion
```

AevaOnline commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+Community Governance Structure
+------------------------------
+
+*Note(Aeva) section added*
```
of course :)

AevaOnline commented on README.md:
```diff
@@ -8,79 +8,47 @@ Open Enclave SDK
-- [Ubuntu 18.04 or 16.04 in simulation mode](docs/GettingStartedDocs/install_oe_sdk-Simulation.md)
-- [Windows Server 2016 with SGX hardware](docs/GettingStartedDocs/install_oe_sdk-Windows.md)
+The Open Enclave SDK is a hardware-agnostic open source library for developing
+applications that run inside Hardware Trusted Execution Environments, also known
```
I realize what the CCC scope is, but do you feel that OE-SDK has applications to run outside of HW TEEs?

gupta-ak replied:
The OE SDK has a "host verification" library that can verify TEE attestation evidence on non-TEEs, so you can technically write apps with the OE SDK that doesn't run as an enclave.

jhand2 commented on README.md:
```diff
@@ -8,79 +8,47 @@ Open Enclave SDK
-[Setting up a Linux Hyper-V VM on Windows with SGX Support](docs/GettingStartedDocs/HyperVLinuxVMSetup.md).
+The OE SDK project [was donated to](https://www.linuxfoundation.org/press-release/2019/08/new-cross-industry-effort-to-advance-computational-trust-and-security-for-next-generation-cloud-and-edge-computing/),
+and is hosted by, the [Confidential Computing Consortium](https://confidentialcomputing.io) (CCC).
+If your company is interested in helping to shape the evolution of confidential
+computing, consider joining the CCC.
```
I'm not sure how to reword this, I just worry that saying "company" here discourages involvement on non-corporate projects

AevaOnline replied:
Projects may be donated to the CCC, regardless of whether they come from a company or not.

However, _membership_ in the CCC costs money, and thus is more geared towards companies.

jhand2 replied:
Makes sense, this is fine then

jhand2 commented on docs/Community/Committers.md:
```diff
@@ -0,0 +1,50 @@
+
+Commits should be fully gated through automated checks, with only a skeleton-crew of folks (a subset of the governance committee) able to directly merge code (and this power is only there for emergency measures, eg. if Prow has failed and needs to be restored).
+
+Ownership should be defined through directory-specific OWNERS files. All actions, such as initiating a test or merging code, should be performed through comments on GitHub that interact with the Prow bot.
```
+1 for this. Managing in a single codeowners file is very painful

AevaOnline commented:
> Good contribution, lots of good things to discuss here.
> I think this PR combines many uncontroversial things with some possibly(?) controversial things.
> If we don't get immediate consensus, it might be helpful to split this long PR into two so that part of it could be merged immediately.

Thanks for the feedback, @dthaler . I can certainly see your point that splitting this into multiple PRs might make it easier to merge. I didn't feel that I had the knowledge to guess that ahead of time, so I opened as one PR, and I'm happy to refactor it.

Do you have any suggestions insofar as where to draw those lines?

achamayou commented on docs/Community/Committers.md:
```diff
@@ -0,0 +1,50 @@
+
+As the project grows, in order to scale the management of the project, it is becoming increasingly necessary to delegate responsibility to domain-owners and improve the community's development workflow through automation. This will be handled through the use of [Prow](https://git.k8s.io/test-infra/prow).
+
+Commits should be fully gated through automated checks, with only a skeleton-crew of folks (a subset of the governance committee) able to directly merge code (and this power is only there for emergency measures, eg. if Prow has failed and needs to be restored).
```
eg is also used: https://dictionary.cambridge.org/dictionary/english/eg

It seems to me that the value of such a comment is sufficiently limited that it is only acceptable as a suggestion block (see below), where the author of the PR just has to click a button to apply the change.

![image](https://user-images.githubusercontent.com/4016369/78187117-c7c02a00-7465-11ea-9738-118cc4e20cb2.png)

dthaler replied:
Thanks, resolving this conversation (I need to learn how to use a suggestion block regardless)

achamayou commented on docs/Community/Committers.md:
```diff
@@ -0,0 +1,50 @@
+
+As the project grows, in order to scale the management of the project, it is becoming increasingly necessary to delegate responsibility to domain-owners and improve the community's development workflow through automation. This will be handled through the use of [Prow](https://git.k8s.io/test-infra/prow).
+
+Commits should be fully gated through automated checks, with only a skeleton-crew of folks (a subset of the governance committee) able to directly merge code (and this power is only there for emergency measures, eg. if Prow has failed and needs to be restored).
```
I think directly here means "without going through the automated checks/process". Line 11-12 does not give reviewers the power to bypass automated checks implemented by Prow, they presumably get discretion to merge only if those checks already pass.

AevaOnline replied:
@dthaler you're absolutely correct. I've addressed this by preserving lines 7 - 14, and changing this section to a description of the current challenges (rather than a prescription for changes). I'll open that in a separate PR.

achamayou commented:
@AevaOnline thank you for proposing this new structure and taking the time to update the doc! I do not have any objections, I've left some minor local comments and suggestions inline.

The only general question I have, and perhaps this PR is too early to ask, is how do we reconcile the strict code ownership property of the SIGs (defined per-directory by Prow, as I understand it), and some of the SIGs mentioned in the document or the discussion such as "Release SIG", "Architecture SIG" or "API SIG"?

Would the release SIG only own packaging scripts? Would it own release branches for patching purposes, across areas normally owned by other SIGs on master?
Would an Architecture SIG own any code at all? How would it be decided that a PR needs to go through the Architecture SIG as well as the SIGs that own the files that the PR modifies? That sounds difficult to automate.

Would there be a default SIG that automatically owns all the code that's not owned by other SIGs?

Apologies if that's too many questions, too early. I certainly don't expect an answer to every one of them, but if there were some sample SIG layouts you could link here, I would be very interested to see how they're structured. That aside, I think having this slightly more formal code ownership mechanism is great, and it's definitely a net improvement over the current situation.

jhand2 commented on docs/Community/Committers.md:
```diff
@@ -0,0 +1,50 @@
+Desired Status
+--------------
+
+As the project grows, in order to scale the management of the project, it is becoming increasingly necessary to delegate responsibility to domain-owners and improve the community's development workflow through automation. This will be handled through the use of [Prow](https://git.k8s.io/test-infra/prow).
```
This probably needs buy-in in a couple ways to become a reality:

1. A proposal of how we should implement prow (what features we want to use, where will it be hosted/infra needs)
2. Someone to actually do the development/integration work

I wonder if it makes sense to split this out into a separate design/issue. Although since it is posed as a desired status instead of current reality I'm also ok if this section stays in this document.

jhand2 replied:
Thanks! We can continue discussion in gh issued.

radhikaj commented on docs/Community/sig-architecture/README.md:
```diff
@@ -0,0 +1,26 @@
+Architecture Special Interest Group
+===================================
+
+SIG-Architecture is responsible for reviewing high level designs, which inform project goals and technical norms.
```
I am not familiar with processes here. Does SIG-Attestation seek consensus from SIG-Architecture? The phrasing here suggests SIG -Architecture must get consensus from other SIGS. I am wondering if the reverse applies as well.

AevaOnline replied:
great question, @radhikaj -- the way I've seen this handled in other open source communities is that a group of technical leaders (representatives, if you will) meet regularly to hash out large architectural questions for the project, agree on "a plan", which is then recorded as a design document.

Each design should include the input of which ever team(s) are ultimately going to do the work and maintain it.

So I think my answer is -- it goes both ways.

radhikaj commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+*Note(Aeva) section added*
+
+The Open Enclave SDK community uses a
+[concensus-seeking](https://en.wikipedia.org/wiki/Consensus-seeking_decision-making)
```
[consensus-seeking](https://en.wikipedia.org/wiki/Consensus-seeking_decision-making)
```suggestion
[concensus-seeking](https://en.wikipedia.org/wiki/Consensus-seeking_decision-making)
```

radhikaj commented on docs/Community/Committers.md:
```diff
@@ -0,0 +1,50 @@
+column describes which technical areas each committer is most interested in, and
+therefore should usually be consulted for changes relating to that area.
+However, it is up to each committer to determine who should review which PR, and
+when to merge it. Remember that a PR must not be merged if a committer objects;
```
Is putting "request changes" against a PR = objection? Are there other ways to object?

AevaOnline replied:
@radhikaj , that was the original wording. I've attempted to clarify in the latest version. Is this better?

radhikaj commented on docs/Community/Committers.md:
```diff
@@ -0,0 +1,50 @@
+
+Commits should be fully gated through automated checks, with only a skeleton-crew of folks (a subset of the governance committee) able to directly merge code (and this power is only there for emergency measures, eg. if Prow has failed and needs to be restored).
+
+Ownership should be defined through directory-specific OWNERS files. All actions, such as initiating a test or merging code, should be performed through comments on GitHub that interact with the Prow bot.
```
There are some problems we have with directory-specific ownership currently. Attestation code for example is not in a specific directory but is in mutliple directories interleaved with regular code.

AevaOnline replied:
Fair enough. Some types of changes may always span across multiple SIGs.

Could the majority of attestation code lives in one location, and changes to other directories (other SIGs) would be more about interfaces, rather than implementation?

jhand2 replied:
This specific example is probably solved by the new attestation plugin design. For other examples I think we could better organize the code to avoid this muddiness.

AevaOnline commented:
@achamayou those are all excellent questions! Any time is reasonable to ask them -- I think it is too early to lock down any answers, but I'm happy to share what's on my mind. Responses inline.

> @AevaOnline thank you for proposing this new structure and taking the time to update the doc! I do not have any objections, I've left some minor local comments and suggestions inline.
> 
> The only general question I have, and perhaps this PR is too early to ask, is how do we reconcile the strict code ownership property of the SIGs (defined per-directory by Prow, as I understand it), and some of the SIGs mentioned in the document or the discussion such as "Release SIG", "Architecture SIG" or "API SIG"?

As both the code and the team grow, the code organization should follow the team organization, as well as the technology alignment.

We already have a directory dedicated to storing design documents -- I would make the sig-architecture leaders responsible for approving changes to `docs/DesignDocs/.` Similarly, there might be a sig-vscode, whose members are responsible for approving changes to `/devex/`

I'm not so sure about sig-api yet ... maybe `/common/`? And I would imagine sig-testing might own `.jenkins/`, `scripts/,` `tests/`, and so on ... but I'm only guessing here.

I imagine that the current Committters would self-organize in what ever way makes sense to the teams.

> Would the release SIG only own packaging scripts? Would it own release branches for patching purposes, across areas normally owned by other SIGs on master?

Ohh... this is something I've seen done differently in different projects. I dunno what makes sense here, but I will hazard a guess that owning the documentation and tooling to generate a release is, at a minimum, something that this SIG would own.

> Would an Architecture SIG own any code at all? How would it be decided that a PR needs to go through the Architecture SIG as well as the SIGs that own the files that the PR modifies? That sounds difficult to automate.
> 

Ah, see above. sig-arch would own the `docs/DesignDocs/` files, and I would expect them to chair the discussions with other sigs to review proposals (bringing in the right people as needed).

> Would there be a default SIG that automatically owns all the code that's not owned by other SIGs?

Basically? but it's called the governance committee, and they would be listed in an OWNERS file in the root directory. In my experience, this should be a super small list of people. For example, kubernetes' top level `OWNERS` has 8 people, though the project has hundreds of committers (listed in other files) and thousands of contributors.

https://github.com/kubernetes/kubernetes/blob/master/OWNERS

> Apologies if that's too many questions, too early. I certainly don't expect an answer to every one of them, but if there were some sample SIG layouts you could link here, I would be very interested to see how they're structured. That aside, I think having this slightly more formal code ownership mechanism is great, and it's definitely a net improvement over the current situation.

This proposal draws very heavily on the kubernetes governance structure. While I've mixed in a lot of learnings from my openstack experience, that is not as useful a comparison. The openstack community management toolset is gerrit+zuul, not github+prow, and that resulted in a significantly different set of processes.

You might be interested to read some things over here ---> https://github.com/kubernetes/community

radhikaj commented:
Thanks for putting thought into how governance would scale.

radhikaj commented on docs/Community/Committers.md:
```diff
@@ -0,0 +1,65 @@
+tools (ie, GitHub itself).
+
+As more developers join the community and bring domain-specific knowledge, there
+is no way to delegate domian ownership, authority or autonomy over changes to
```
```suggestion
is no way to delegate domain ownership, authority or autonomy over changes to
```

radhikaj commented on docs/Community/Committers.md:
```diff
@@ -0,0 +1,65 @@
+Current Expectations
+--------------------
+
+However, it is up to each committer to determine who should review which PR, and
```
I am wondering if I as a committer open a PR, can I decide that only committer A and committer B review it? If committer C reviews it despite not being assigned, can C request changes? I understand that this is probably what was in the old docs, so I am just leaving this in so I understand the answer and dont ask for this to be addressed in the current PR.

AevaOnline replied:
Yeah, that's what was in the old doc, and this situation crossed my mind as well.

I would propose that approval from at least one SIG lead other than the authori is required to merge code, and approval from two or more is preferred (especially for larger changes)

achamayou commented:
@AevaOnline thank you for the answers and for the pointers to the way this is done in kubernetes, it's very helpful!

It sounds like SIG Architecture is almost a hybrid SIG/WG, and the decision of when it should be involved is essentially the same as the decision for when to produce a Design Doc at the moment. Or at least that's how I think of it.

AevaOnline commented:
> a hybrid SIG/WG

@achamayou I feel the tension between these two definitions here as well...

> the decision of when it should be involved is essentially the same as the decision for when to produce a Design Doc at the moment.

That's how I look at it as well.

I lean towards the SIG side since there are in-tree deliverables (namely design docs) today, but if the design docs were housed in a separate repo, I would lean more towards the WG side. For what it's worth, in the large projects I've worked in, design docs have always ended up in their own repo. (I'm not suggesting design docs need to split out, just providing this for context as to how I'm thinking about SIG--WG)

achamayou commented:
@AevaOnline I don't think this is a problem, I'm just checking my understanding, and it still sounds a net improvement. I am also not sure the design docs warrant a separate repo at this point, but this may change of course.

CodeMonkeyLeet commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+conversations (for those who work in the same building, or bump into each other
+at a conference). The outcome of such discussions should be considered
+non-binding until the information is shared through an open medium and relevant
+stake-holders have opportunity to contribute to the discussion.
```
```suggestion
stake-holders have the opportunity to contribute to the discussion.
```

CodeMonkeyLeet commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+At present, we maintain a list of [Project Committers](Committers.md) along with
+descriptions of their knowledge areas.
+
+This will be updated and replaced with a more scaleable approach to
```
```suggestion
This will be updated and replaced with a more scalable approach to
```

CodeMonkeyLeet commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+than the contributor, and have no significant objections from other SIG leads
+within the same SIG.
+
+If a change requires the approval of multiple SIGs (e.g., because it affects
```
Tangent: Does this imply the adoption of new tooling that supports this semantic? (e.g. prow & LGTM)

AevaOnline replied:
If bors does not support this requirement, then yes

As I understand it, Prow was build out of the needs of another distributed, multi-party, github-driven workflow, very similar to (if not exactly the same as) what I believe OpenEnclave will need as we grow.

CodeMonkeyLeet commented on docs/Community/Committers.md:
```diff
@@ -0,0 +1,65 @@
+--------------------
+
+However, it is up to each committer to determine who should review which PR, and
+when to merge it. It is up to each committer to decide [when to accept a contribution](Governance.md#Accepting-Contributions)]
```
```suggestion
when to merge it. It is up to each committer to decide [when to accept a contribution](Governance.md#Accepting-Contributions).
```

CodeMonkeyLeet commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+If a change requires the approval of multiple SIGs (e.g., because it affects
+code in both SIG domains), then approval from both SIGs is required.
+
+If a change only requires one approval from one SIG, and leads of another SIG
```
The Committers.md still has old text that points here, but still says:
> However, it is up to each committer to determine who should review which PR, and
> when to merge it. It is up to each committer to decide [when to accept a contribution](Governance.md#Accepting-Contributions)

A couple of clarifications:
* Is the implication that every committer is a member of a SIG? Is a distinction being made between a contributor as a SIG member vs. a SIG lead?
* If not, what would be the relationship between the approval of individual committer and a SIG member? 
   * e.g. in our current 2 approvals required policy, it could be any 1 committer plus a SIG lead, but not 2 non-SIG committers?

AevaOnline replied:
This is an area that I would like to discuss and clarify more. Agree that it's not clear right now, even with the draft changes I propose.

Here's my thinking:
- remove the idea of "committers" that is present today, replacing that with "approvers" and "reviewers"
- using automation (eg, prow & lgtm, as mentioned above), distribute fine-grained rights management through OWNERS files in each directory
- "reviewers" are able to leave "lgtm" comments, which the bots aggregate to manage github tags and gating tests
- "approvers" are able to leave "approve" comments, which, when there are required # of approvals from the right people, then trigger a merge to be performed by the bots
- reduce # of people who actually have write access (eg, can "force push") to the repo to the absolute bare minimum of 3 - 4 people. Less than that risks project continuity if there's an accident

See https://prow.k8s.io/command-help for listing of some of the commands k8s uses. Commands are fully customizable as well.

Following this approach, a commit would need to have two "lgtm" and one "approve" comments, from people who have the correct permissions to review and approve changes to that portion of the code base. This is only a proposal -- needing three sets of eyes on a change may be too many for the current team size. My main point is that the tooling would help us here by introducing an explicit and transparent review structure, delegating both authority and autonomy to domain-owners, and thereby allowing the review workload to spread out more.

CodeMonkeyLeet replied:
I like the structure of that, thanks for elaborating on the thinking there! I'm primarily looking for clarity on what changes in processes between what exists today and how to get people to transition to different practices described in the new text; a couple of additional clarifications and thoughts:

* I mentally map "approvers" to "committers" today in the sense that these are the individuals identified in the CODEOWNERS file today that would map to OWNERS in prow. 
    * Is "reviewers" a special category in that case? Prow describes them as any collaborators on the project, and so in terms of access rights, these would just be equivalent to general contributors in the OE project? i.e. there is no special process to identify or approve individuals as being "reviewers".
    * My impression (please correct) is the difference between "approvers" and "committers" is "approvers" do not have the same access privileges as implied by "committers" today:
        * "approvers" do not have elevated project privileges like write or force push. Their authority to merge is controlled through automation (i.e. prow)
        * Do "approvers" get heightened privileges in configuring prow as well? (i.e. committers today through their privileges on the project also have elevated privileges on bors)
        * If the idea is to restrict project privileges to a handful of individuals, we should probably also describe the process for selecting them since it sounds like this is likely a subset of the CGC members? That could probably be deferred to a future CGC meeting discussion. 
* Yeah, 3 individuals per PR is probably more than the current team size can keep up with right now, given the code velocity to date. I would suggest we keep the current requirement of two approvals or an LGTM and an approval until the project scales up. This is likely even more true as we fragment review responsibilities to individual SIGs.

AevaOnline replied:
> "approvers" do not have elevated project privileges like write or force push. Their authority to merge is controlled through automation (i.e. prow)

Correct

> Do "approvers" get heightened privileges in configuring prow as well? (i.e. committers today through their privileges on the project also have elevated privileges on bors)

The list of approvers, being determined by an OWNERS file in a given directory, becomes self-governing through Prow. Those approvers (and anyone who supercedes them) become the folks who can change the list of approvers within that OWNERs file.

If prow configuration itself were managed through code in a directory, then the approvers listed in the OWNERS file in _that_ directory would be the list of people who can control prow's configuration.

In this way, many people may be able to approve project changes _without_ also being able to control the prow configuration. The infra team would control the infrastructure. 

> If the idea is to restrict project privileges to a handful of individuals, we should probably also describe the process for selecting them since it sounds like this is likely a subset of the CGC members? That could probably be deferred to a future CGC meeting discussion.

Agreed. Worthwhile discussion for a future CGC meeting, and/or a thread on the CGC ML.

As the project grows, gains more contributors and more attention, keeping the set of people who retain project privileges (eg, can force push) to a very small set increases in importance. This is the set of people who are trusted with the power to break the project history (and who, clearly, would not do that unless absolutely necessary).

I'm not concerned with anyone on the current list; rather, my concern is that, with the process that is in place today, I don't see a way to expand the approver list without expanding the committer list, or to delegate domain-specific approval rights.

CodeMonkeyLeet commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+
+If a change only requires one approval from one SIG, and leads of another SIG
+object to it, a discussion should be invited and given reasonable time to move
+towards a consensus; overruling the objection of another SIG lead should only be
```
* Charter.md still has old text that suggests that the CGC is still an arbitration body of last resort for _technical_ decisions as well.

   >This Committee is not intended to make every technical decision, as those should
generally be made by agreement among committers as PRs are reviewed and merged.
Where disagreements take place and need further resolution, those can be brought
up with the Committee as part of its responsibility to maintain the project's
community health. Otherwise technical decisions are left to the active
committers (by virtue of the liberal contribution policy).

   Is that still accurate in the new governance structure, or should that be revised and made clear that SIGs are responsible for arbitration of technical issues?

* Is there an expectation that all SIGs are co-equal in this regard? For example, I can imagine that SIG-architecture will need to drive decisions across SIGs regularly, and how do those relationships work?

* Other projects have the notion of a technical steering committee, which do serve as a final body for technical arbitration. In the proposed taxonomy, would that be like a kind of permanent WG with representation from all SIGs? Is that a useful concept?

AevaOnline replied:
Great question ... yeah, I could revise this section further.

Every other projects I've been involved in had a "technical committee" of some variety, whether it was called the TC / SC / TAC / TOC, or the project was led by a BDFL (*). That committee (or individual) also determined the governance structure, resolved disputes between projects or SIGs, and some of them also put effort into the community health.

I don't know that unifying both technical and community leadership in one body is better than separating them, especially at an early stage. My suggestion would be to keep them unified right now. Perhaps that committee is made up of the SIG leads? *shrug*

Oh, and Kubernetes also has a "conduct committee", though that is really an escalation point for serious conduct issues or consent incidents, which local leads often aren't trained to handle.

(*) acronym decoder:
- TC: openstack [technical committee](https://governance.openstack.org/tc/)
- SC: kubernetes [steering committee](https://github.com/kubernetes/steering)
- TOC: CNCF [technical oversight committee](https://github.com/cncf/toc)
- TAC: CCC [technical advisory committee](https://github.com/confidential-computing/governance/blob/master/scoping.md) ... not the greatest docs, but also a much younger body
- BDFL: benevolent dictator for life (eg., Linus Torvalds, Guido van Rossum, ...)

AevaOnline replied:
As an aside, in this proposal I am seeking to address current development and review scaling challenges by proposing a system for delegating responsibility and improving the github workflow. I did not want to include a proposal for a new structure for the CGC, or for a technical governance committee, because I know that I don't know enough about that layer yet.

When I do my next iteration of this changeset, I'll see if there are ways I can scope the changes more narrowly.

CodeMonkeyLeet replied:
> I did not want to include a proposal for a new structure for the CGC, or for a technical governance committee, because I know that I don't know enough about that layer yet.

I agree that it is not necessary to restructure the CGC or its responsibilities towards/away from technical/steering committees used in other open source projects as part of this PR. My last bullet was really a guess at whether you were suggesting going down that road. Returning to the scoped concerns:

> - Is that still accurate in the new governance structure, or should that be revised and made clear that SIGs are responsible for arbitration of technical issues?

It sounds like the intention is that if SIGs do not come to an agreement, then yes, the CGC would be responsible for arbitrating technical disagreements for now. If so, that should probably be stated in the text here; The statement that SIG leads should not overrule each other does not currently indicate where the process bottoms out, and I think that would be helpful to include.

> - Is there an expectation that all SIGs are co-equal in this regard? For example, I can imagine that SIG-architecture will need to drive decisions across SIGs regularly, and how do those relationships work?

I'm not sure I have clarity on this question. Elsewhere in the PR discussions, it seems to be suggested that SIGs are really divided along lines of directory ownership for automation purposes, so the SIG-architecture would be defined by its ownership of docs/DesignDocs. I think some additional elaboration about how the SIGs relate to one another would be helpful here. 

For example, if we don't want to decide to establish a technical steering committee at this point, it seems equally useful to state that SIG leads from other SIGs should also be included in SIG-architecture because it's pretty much everyone's responsibility to review design documents in docs/DesignDocs.

AevaOnline replied:
> For example, if we don't want to decide to establish a technical steering committee at this point, it seems equally useful to state that SIG leads from other SIGs should also be included in SIG-architecture because it's pretty much everyone's responsibility to review design documents in docs/DesignDocs.

Hmm... what do you think about a structure where all sig leads are "reviewers" in the docs/DesignDocs folder, and the sig-arch lead(s) function is to facilitate the discussion to reach consensus before approving any change? Basically, sig-arch lead is a meeting chair that makes sure all the right people are included in the review process for any given doc before approving it.

Does that help clarify my thinking on this? 

It may be that we need to establish a technical steering committee at the same time as SIGs, and that's OK. I think your points about having a clear resolution path for cross-sig disagreements is completely valid and I'm not sure of the best way to address it today.

CodeMonkeyLeet replied:
>Hmm... what do you think about a structure where all sig leads are "reviewers" in the docs/DesignDocs folder, and the sig-arch lead(s) function is to facilitate the discussion to reach consensus before approving any change? Basically, sig-arch lead is a meeting chair that makes sure all the right people are included in the review process for any given doc before approving it.

Yeah, I think that does help clarify the approach and would be helpful to include in the docs. Thanks for working through my questions! 

Regarding establishing a technical steering committee at the same time as the SIGs, I don't have a strong opinion either way; I think what I'm mostly looking for is something in the governance that describes the expectations around how SIGs coordinate and I think a TSC is kind of a common way to deal with that. I think this discussion provides the kind of touchpoints I was looking for either way:

- I think it's really valuable if a process provides guidance for resolving disagreements and deadlocks as a backstop, so an escalation to CGC as previously proposed does address that.
- I think it's better still if governance builds in coordination mechanisms to reduce the occurrence of such issues. For example, I think of the clarification you provided for how SIG leads review the DesignDocs folder as being valuable in that regard.

AevaOnline commented on docs/Community/Committers.md:
```diff
@@ -0,0 +1,65 @@
+therefore should usually be consulted for changes relating to that area.
+
+
+| Name                  | GitHub Alias        | Area                           |
```
Great question, @CodeMonkeyLeet ! For a GitHub-based workflow, I would strongly recommend [Prow](https://github.com/kubernetes/test-infra/tree/master/prow). Take a look here for a combination of war stories and hope from the guy who built sig-testing in Kubernetes  - https://youtu.be/BsIC7gPkH5M?t=1033

And given the move to SIGs, yes, I would suggest that we align committer areas to sigs to directory structure (where possible).

AevaOnline replied:
Oh, here's a more recent / updated edition of his talk - https://www.youtube.com/watch?v=7-_O41W3FRU

mingweishih replied:
Prow looks great. Thanks @AevaOnline for the pointer.

AevaOnline commented on docs/Community/Communication.md:
```diff
@@ -0,0 +1,22 @@
+Chat Systems
+------------
+
+TBD
```
That seems reasonable to me. Would you like to make a suggestion?

Admittedly, I don't know where the gitter instance is or how to access it ...

CodeMonkeyLeet replied:
It's already a badge on the front page README.md: https://gitter.im/openenclave/community

AevaOnline replied:
ah - thanks! updating and marking as resolved

kuanggang-git commented:
Can one of the admins verify this patch?

andschwa commented on docs/Community/conduct/IncidentResponseGuidelines.md:
```diff
@@ -0,0 +1,7 @@
+This document should outline the policy and guidelines which the recipient of a
+Conduct Incident report will follow.
+
+TODO
```
What a big TODO 😅 There is some helpful guidance in https://www.contributor-covenant.org/faq/ and in https://opensource.guide/code-of-conduct/

andschwa commented on docs/Community/conduct/README.md:
```diff
@@ -0,0 +1,18 @@
+down our general views on how conflict between community members
+[should be handled](ConflictPolicy.md).
+
+Incidents may be reported to the
```
```suggestion
Incidents should be reported to the
```

I want somehow to encourage folks to submit reports, but I'm not sure how to do that.

andschwa commented on docs/Community/governance/README.md:
```diff
@@ -0,0 +1,40 @@
+
+The [charter](charter.md) defines the scope and processes by which the CGC operates.
+
+Meetings: public CGC meetings are held on Wednesdays (bi-weekly) (*link TBD*)
```
The meetings are not currently public, only the minutes. We need space to discuss things confidentially. We may want to have a public meeting (I believe this was written as a suggestion in these documents previous iteration) and so I am in support of having regular, public community meeting.

dthaler replied:
Is there anything (other than manually watching who's in the meeting) that enforces that it's public today?  The minutes link below at https://aka.ms/openenclave/cgc contains the link to the Zoom meeting, so the link is currently public if the minutes are public, and the Zoom meeting currently does not require a password or anything.

andschwa replied:
Let me rephrase, we have not invited the public to these meetings. Sure if people start randomly showing up perhaps we need to add a password etc. but regardless, the committee has chosen not to make these meetings public.

AevaOnline replied:
updated the doc to reflect today's reality (that we don't really have public meetings)

andschwa commented on docs/Community/governance/README.md:
```diff
@@ -0,0 +1,40 @@
+
+Meetings: public CGC meetings are held on Wednesdays (bi-weekly) (*link TBD*)
+
+Minutes: aka.ms/openenclave/cgc
```
```suggestion
Minutes: https://aka.ms/openenclave/cgc
```

(So it renders as a link.)

andschwa commented on docs/Community/governance/README.md:
```diff
@@ -0,0 +1,40 @@
+How to raise a topic to the CGC
+-------------
+
+TODO
```
Ties into a comment of mine above where we may want to have an email which anyone can send messages to the CGC (so like oesdkcgc but public).

andschwa commented on docs/Community/governance/charter.md:
```diff
@@ -0,0 +1,57 @@
+and work to resolve any negative issues.
+
+In order to maintain a healthy developer community, it is recommended that the
+Committee also host a regular public community meeting. This meeting should be
```
Oh here's that suggestion!

andschwa commented on docs/Community/governance/charter.md:
```diff
@@ -0,0 +1,57 @@
+Governance Committee Charter
```
Did any changes take place to this document? GitHub is not recognizing it as a rename, so I can't tell if it was renamed _and_ modified.

AevaOnline replied:
At the time you asked this question, I think I had made some minor changes. But as of the latest revision, I have made considerable changes. One thing I'd like to call out, which was in the original and remains in this version, is that the CGC charter explicitly says that it is not the place for technical decisions to be made. I've attempted to clarify that _a little bit_ in this draft, by referencing SIGs and WGs.

dthaler commented on docs/Community/CodeOfConduct.md:
```diff
@@ -0,0 +1,11 @@
+following a set of [reponse guidelines](conduct/IncidentResponseGuidelines.md)
+for any reports.
+
+Instances of abusive, harassing, or otherwise unacceptable behavior may be
```
My opinion: As noted in PR #2744 I think anyone ought to be able to email the oesdkcgc ML

AevaOnline commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+approach to governance at all levels. To support a growing project, we have
+adopted a loose hierarchy wherein domain-specific decisions are handled within
+each domain, or Special Interest Group (**SIG**), cross-domain work is enabled
+through Working Groups (**WG**), and a Community Governance Committee (**CGC**)
```
summarizing some discussions, it seems as though formalizing _that_ these structures can exist (SIGs and WGs) is really a matter of defining _how_ one might come to be and _what_ its responsibilities would be. That meta-process was lacking, and discussing through what some of the initial SIGs might be has also helped me better understand what meta-structure is missing from the current proposal.

examples of questions I've fielded several times:
- how many SIGs is too many?
- does every SIG need a weekly meeting?
- is every SIG responsible for code?
- does CGC determine SIG membership?

CodeMonkeyLeet replied:
I found the new text to be quite helpful with regards to these questions, thanks!

AevaOnline commented on docs/Community/conduct/ConflictPolicy.md:
```diff
@@ -0,0 +1,26 @@
+yield positive results. However, it is never okay to be disrespectful or to engage
+in behavior that violates the project’s [Code of Conduct](../../CODE_OF_CONDUCT.md).
+
+If you see someone violating the code of conduct, you are encouraged to address
```
I whole-heartedly agree with this feedback. Do you think that should be rolled in to this gov patch, or fixed in a separate patch?

AevaOnline commented on docs/Community/conduct/ConflictPolicy.md:
```diff
@@ -0,0 +1,26 @@
+provide them an opportunity to discuss it before any action is taken. The identity
+of the reporter will be omitted from the details of the report supplied to the accused.
+In potentially harmful situations, such as ongoing harassment or threats to anyone’s
+safety, we may take action without notice.
```
I'd love to see that as well.

a transparency report sounds like enough of a "deliverable" that it should be more clearly documented whose responsibility it is to track incidents and generate the report. Do you think that's a CGC goal, or a CoCC goal (if/when such a body is created)?

AevaOnline commented on docs/GettingStartedDocs/README.md:
```diff
@@ -0,0 +1,93 @@
+
+Open Enclave provides a consistent abstraction across different hardware Enclave
+platforms. If you're just getting started, you can compile and test your apps
+without an enclave (**TODO: add link to docs for this**), or you can use one of
```
Ah. I suppose I simply missed it. Removing the TODO

```suggestion
without an enclave, or you can use one of
```

AevaOnline commented on docs/Community/Communication.md:
```diff
@@ -0,0 +1,22 @@
+
+The community currently relies on two mailing lists:
+
+1. "oesdk at lists dot confidentialcomputing dot io" -- this is the primary mailing list for discussions and announcements.
```
done

AevaOnline commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,137 @@
+approximately how the community operates and the principles which infuse our
+culture today.
+
+It should be edited and revised from time to time.
```
Thanks! :)

Closing as there is no action to take.

AevaOnline commented:
In addition to incorporating several folks' editorial corrections (thank you all!), I've made the following updates based on discussion in the last CGC meeting:
- added draft for a sig charter template -- docs/Community/governance/sig-charter-template.md
- added a first pass at creating a charter for sig-release -- docs/Community/sig-release/charter.md
- added OWNERS files for the list of SIGs which were discussed within the CGC meeting
- took a guess at the current membership of each SIG, based on the current Committers list

Also updated the following files based on notes and my recollection of the CGC discussion:
- docs/Community/Governance.md
- docs/Community/governance/charter.md

AevaOnline commented:
@gupta-ak  - thanks for the review. 

Let me see if I can answer your questions. I hope my latest updates also help to answers these :) 

>     * What would be the process of joining a SIG?

That is up to each SIG, based on the responsibilities and process which that sig defines in its charter. There's a spot in the `sig-charter-template.md` file now where a sig should fill out this information when the sig is created.
 
>     * How do we create a new SIG?

The CGC should be empowered to create SIG / WG as needed. I've updated `docs/Community/governance/charter.md` to clarify this, but let me know if you think it would benefit from more explanation.

>     * What is the relationship between code ownership and the SIG? How would that be enforced by code tooling? For example, for attestation, there are a bunch of attestation related code spread throughout the `host`, `enclave` and `common` directories. Would all these files be owned by the attestation SIG?

Today, I don't think we can enforce this, just like we can't enforce the current list of "areas" in Committers.md ... 

That said, in other projects, this is enforced through OWNERS files which can be placed within *any* directory.  In this way, for example, if attestation code lived in both `host/measure/` and `enclave/measure/` directories, then the OWNERS file in those directories could list the members of sig-attestation.

AFAIK, Prow already supports this functionality, including through the definition of macros which would allow (in the example above) both of those OWNERS files to reference a "sig-attestation-approvers" metalist, which might be defined in `docs/Community/sig-attestation/OWNERS`. It's entirely possible other solutions exist as well.

CodeMonkeyLeet commented on docs/Community/governance/OWNERS:
```diff
@@ -0,0 +1,7 @@
+chair:
+
+reviewers:
```
As a general comment, can we order all of the OWNERS in responsibility level as described in the sig-charter-template.md? It also provides the opportunity drop empty categories from files that don't need them yet (e.g. emeritus, member)

AevaOnline replied:
done

CodeMonkeyLeet commented on docs/Community/governance/sig-charter-template.md:
```diff
@@ -0,0 +1,82 @@
+### Chair
+
+- Role: Chair of a SIG or WG
+- Responsibilities
```
For SIGs, are the responsibilities cumulative as you move from Reviewer to Approver to Chair? I would assume so, but the chair responsibilities seem pretty distinct from the other two.

AevaOnline replied:
I think that's up to the community // that SIG to decide.

in my experience, the privileges are cumulative but the responsibilities are not. For instance, a chair of a busy SIG may not have time to do as many reviews as a reviewer, since the chair's role is more organizational and less tactical.

CodeMonkeyLeet commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,161 @@
+SIGs will generally accept changes that improve the project and that align with the
+[Open Enclave roadmap](https://github.com/openenclave/openenclave/projects).
+
+All contributions are expected to be reviewed by at least one *Reviewer* or
```
This is a nice clarification that still leaves room for flexibility, thanks for the change here.

CodeMonkeyLeet commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,161 @@
+through Working Groups (**WG**), and a Community Governance Committee (**CGC**)
+meets to review the overall structure, ensuring the health of the project.
+
+_NOTE: In lieu of a "technical steering committee", the Architecture SIG should
```
I think this note can be removed from the governance section as you've also added similar  text in the technical discussions section below, which seems like a better fit for the topic. (maybe bump that last paragraph up to the front if it's not visible enough).

CodeMonkeyLeet commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,161 @@
+Each group (SIGs, WGs, and CGC) shall maintain documentation in this repository.
+Creating a new SIG or WG requires the approval of the CGC, and is formally
+created by the merging of a "charter" document. A SIG is generally responsible
+for the delivery of a technical artefact (code, release binaries, operation of
```
Use the American spelling ...
```suggestion
for the delivery of a technical artifact (code, release binaries, operation of
```

CodeMonkeyLeet commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,161 @@
+When a change affects code owned by multiple SIGs, it should be coordinated with
+the respective SIGs. If long-term collaboration is anticipated, cross-SIG
+efforts can be coordinated by establishing a Working Group (WG), which is given
+a charter and which reports progress back to the relevant SIGs and/or the CGC.
```
Actually, maybe here is the right place to bring up the SIG architecture (lines 102-105); it seems to deserve some special mention because it's effectively a permanent WG across all technical SIGs, that also produces artifacts (design docs etc.)

radhikaj commented on docs/Community/governance/sig-charter-template.md:
```diff
@@ -0,0 +1,82 @@
+
+## Process & Management
+
+THIS SIG meets at least THIS OFTEN.
```
SIG-Release does not meet at all. Are SIGs required to meet?

dthaler replied:
I think we previously said no, SIGs are not required to ever meet

AevaOnline replied:
I believe I'd said (and certainly intended to say) that SIGs do not have to meet. I can see how that's not clear in the charter-template.

That being the case, I also believe that there is value in having a transparent process, so a SIG which does not meet should endeavour to have a clearly communicated process via some other means. 

I'll think about the best way to indicate this flexibility in the charter....

radhikaj commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,161 @@
+charter.
+
+If a change requires the approval of multiple SIGs (e.g., because it affects
+code in both SIG domains), then approval from both SIGs is required.
```
So changes to the public API surface, if it related to attestation, needs approval from SIG-Architecture?

dthaler replied:
My view: yes, since SIG-Arch is responsible for checking for adherence to the API guidelines and consistency across the full public API surface area

radhikaj commented on docs/Community/sig-release/charter.md:
```diff
@@ -0,0 +1,47 @@
+
+SIG-Release meets only as needed. Meetings are generally more frequent in the run-up to a release.
+
+Meetings are held at PLACE, with AGENDA/MINUTES available or not. Meetings are OPEN/CLOSED.
```
@AevaOnline Sig-release may not need to meet. Most of the work is on Github. It does need the CI/CD team involvement to generate and validate packages. To that end, I would like @BRMcLaren to be included in this SIG as well.

AevaOnline replied:
@radhikaj is there a check-in or sign-off process for cutting a release? An informal discussion that happens between you and @BRMcLaren when generating and validating the packages? If so, I'd say that's a "meeting". 

It doesn't need to be a regularly scheduled meeting, but in the interest of building community engagement, I think it should be documented, communicated, and open to the community.

An example of this might be:

> There is usually just one release meeting held between the sig-release chair and the sig-cicd chair during a given release cycle, and it is usually held the week before the release. As 99% of the work happens on github, this meeting is only a brief touch point to validate expectations and confirm that the builds are ready to be cut and finalized. The meeting will be announced at least three days ahead of time on the mailing list.

If there is truly no need for any discussion about the release readiness, and all coordination happens on github, that's also perfectly in keeping with the Principles,  and I would just ask for it to be documented in this charter so that the process is transparent. Something like:

> SIG-Release does not hold any meetings. All preparation for releases happens via github issues, tags, and tracking boards. (insert link to document of the process)

radhikaj replied:
The process for release is - SIG-testing generates RC packages. We do an RC release and send out an email to the DL requesting feedback. The OE SDK core team tests the OE SDK packages manually. SIG-Testing generates final release packages once the release manager says the release branch is in its final state. We do another round of manual testing. We then make a final release.

radhikaj commented on docs/Community/sig-release/charter.md:
```diff
@@ -0,0 +1,47 @@
+
+The current SIG Chair is: Radhika.
+
+**Note (Aeva)** -- Proposed succession process here **please review**
```
+1 for above.

AevaOnline replied:
well-defined chain of succession is certainly a reasonable approach :) 

@radhikaj as the sig-release chair, you get to define that plan. This was merely me making a proposal based on how I've run projects before. Please customize to your liking -- all that matters is a) that there is a plan, b) that it's documented, and c) that it works when we need it to ;)

dthaler commented on docs/Community/sig-attestation/README.md:
```diff
@@ -0,0 +1,26 @@
+
+SIG-Attestation is responsible for code, plugins, testing, and other works
+related to or required for attestation and integration with third-party /
+external attestation services.
```
I think SIG-Attestation should also be responsible for tracking OESDK's alignment with attestation standards
(IETF RATS, TCG, etc.) and would like to see that stated explicitly.   I personally spend a portion of my time doing just that.

AevaOnline replied:
added

radhikaj commented on docs/Community/sig-architecture/OWNERS:
```diff
@@ -0,0 +1,17 @@
+  - sig-tools-chair
+  - sig-release-chair
+  - sig-sgx-chair
+  - sig-optee-chair
```
After reviewing the docs, I think it makes sense to have a SIG-API.

dthaler replied:
Maybe document why website/docs are not on the list?

AevaOnline replied:
added sig-api (empty list for now) and a note about sig-website

dthaler commented on docs/Community/sig-optee/OWNERS:
```diff
@@ -0,0 +1,11 @@
+chair:
+  - dthaler
```
I'd say someone else (maybe Hernan) would be better at this one

AevaOnline replied:
I've removed all the chair appointments from this pass, so that SIGs can properly self-organize :)

radhikaj commented on docs/Community/sig-sgx/OWNERS:
```diff
@@ -0,0 +1,17 @@
+chair:
+  - TBD
+
+reviewers:
```
I think our original list of committers and areas of ownership are outdated. This is not specific to SGX. We should review those areas and then perhaps figure out the list of approvers in various areas. That can be done after this doc is merged and is not a blocker, but just something the CGC needs to do.

radhikaj commented on docs/Community/sig-architecture/OWNERS:
```diff
@@ -0,0 +1,17 @@
+# reviews should be solicited from all SIG Chairs
+# NOTE(Aeva) missing domains: API, CCF, Windows
+reviewers:
+  - sig-attestation-chair
```
sid-cicd?

dthaler commented on docs/Community/sig-sgx/OWNERS:
```diff
@@ -0,0 +1,17 @@
+reviewers:
+
+approvers:
+  - anakrish
```
It would be good if Intel had representation in SIG-SGX, or even as chair

achamayou commented on docs/Community/sig-architecture/OWNERS:
```diff
@@ -0,0 +1,17 @@
+  - TBD
+
+# reviews should be solicited from all SIG Chairs
+# NOTE(Aeva) missing domains: API, CCF, Windows
```
Agreed, I don’t think it’s necessary to have a specific SIG for CCF. I would prefer to get involved in SIG Architecture or API if that becomes separate.

AevaOnline replied:
interesting. good points. I was just copying info from the old file :)

it is sounding like a sig-api might be needed at some point...

dthaler commented on docs/Community/Committers.md:
```diff
@@ -0,0 +1,70 @@
+As more developers join the community and bring domain-specific knowledge, there
+is no way to delegate domain ownership, authority or autonomy over changes to
+code. And as the community itself grows, the current communication channels do
+not support rapid iteratation, or facilitate remote and open collaboration.
```
```suggestion
not support rapid iteration, or facilitate remote and open collaboration.
```

dthaler commented on docs/Community/Governance.md:
```diff
@@ -0,0 +1,161 @@
+
+- **Chair**: the primary point of contact for a SIG or WG, responsible for the processes within the SIG; a tie-breaker in votes within the SIG, if needed
+- **Approver**: members of the SIG who are empowered to approve code changes
+- **Reviewer**: members of the SIG whose knowledge is respected by Committers, and whom are frequently asked to provide code reviews for patches
```
Is there a difference between "Committer" as used in this line, and "Approver" as used in the previous line?  They sound synonymous, and if so can we just use one term throughout?

AevaOnline replied:
```suggestion
- **Reviewer**: members of the SIG whose knowledge is respected by Approvers, and whom are frequently asked to provide code reviews for patches
```

jhand2 commented on docs/Community/sig-attestation/OWNERS:
```diff
@@ -0,0 +1,13 @@
+chair:
+  - soccerGB
```
I also think @shnwc should be an approver, although we don't have the current tooling to support this as he is not a committer

AevaOnline commented on docs/Community/sig-release/charter.md:
```diff
@@ -0,0 +1,47 @@
+SIG Release Charter
```
that's an interesting question -- I didn't intend to _change_ the release (approval) process in this patch. 

I'll leave the docs in their current state, even if as you point out it is inconsistent, so that it can be discussed and corrected without me having to guess at the process

AevaOnline commented:
As per CGC meeting and agreement on April 24th, I have updated this PR and will now merge it. I am noting this since a self-approval is uncommon, and was explicitly requested in the meeting.

Additional changes e.g., to establish the SIG memberships and chairs, shall be done in subsequent PRs.

AevaOnline commented:
bors r+

AevaOnline commented:
bors retry

AevaOnline commented:
bors retry
